Adds mathematics page (#21)
NUbots/NUbook: src/book/02-projects/01-nubots/12-mathematics.mdx

A transformation matrix in three dimensions is a 4x4 matrix containing a **rotation** component and a **translation** component.

If we have some rotation $\mathbf{R}_a^b$ and a translation component $\mathbf{u}^b$, then we have a transformation matrix that transforms from space $a$ to space $b$:

$$
\mathbf{H}_a^b = \begin{bmatrix}
\mathbf{R}_a^b & \mathbf{u}_{BA}^b \\
\mathbf{0} & 1
\end{bmatrix}
$$

This is written in the NUbots codebase as `Hba`.

The vector $\mathbf{u}_{BA}^b$ is the vector from A (the origin of space a) to B (the origin of space b) in b space, which would be written in the codebase as `rBAb`.

The reason for this is that if we apply the transform $H^b_a$ to some point $\rho^a$, then $\rho^a$ will first be rotated by $R_a^b$, giving $\rho^b$. $\rho^b$ will then be translated by some vector according to the transformation matrix. Since $\rho^b$ is now rotated in the $b$ coordinate system, the translation vector must also be in this $b$ coordinate system too. Thus, the translation vector is $\mathbf{u}_{BA}^b$.

Transformation matrices left multiply with vectors to get a vector in the new space.

$$
\mathbf{v}^b = \mathbf{H}^b_a \cdot \mathbf{v}^a
$$

> This is written in the codebase as `vCDb = Hba * vCDa`, if $\mathbf{v}$ represents a vector from point D to point C.

Reviewer: We do in some places use `v` rather than `r`, usually when we are representing a velocity (see https://github.com/NUbots/NUbots/blob/master/module/platform/darwin/SensorFilter/src/MotionModel.h#L43 and https://github.com/NUbots/NUbots/blob/biddulph/eigenify_platform/module/platform/darwin/SensorFilter/src/MotionModel.h#L52 for a comparison, `rTWw` vs `vTw`).

I changed `rCDb = Hba * rCDa` to `vCDb = Hba * vCDa` because you have `\mathbf{v}^b` and `\mathbf{v}^a` in the equation above. But it might be better if we changed

`\mathbf{v}^b = \mathbf{H}^b_a \cdot \mathbf{v}^a` to `\mathbf{r}^{b}_{BA} = \mathbf{H}^b_a \cdot \mathbf{r}^{a}_{BA}`

Author: Yeah I think changing the name of the vector to `r` in the equations would work better then.

$$
\mathbf{r}^a = \begin{bmatrix}
r_1^a \\
r_2^a \\
r_3^a \end{bmatrix}
$$

To make computations such as inner product or cross product with coordinates, they must be relative to the same basis.

## Rotations

To describe the motion of the robot, we need a **reference frame** and a **coordinate system**.

The reference frame is the physical rigid body and the coordinate system is a mathematical concept, which is represented as a basis.

**Rotations** are defined by a 3x3 matrix. If we have a vector $\mathbf{r}^a$, that is the vector $r$ in the coordinate system $a$, the rotation

$$
\begin{bmatrix}
\cos(\psi) & -\sin(\psi) & 0 \\
\sin(\psi) & \cos(\psi) & 0 \\
0 & 0 & 1 \end{bmatrix}
$$

will rotate $\mathbf{r}^a$ by $\psi$ around the coordinate axis $\mathbf{a_3}$.

The orientation of the basis $b$ relative to $a$ can be described by three consecutive rotations about the main axis. There are 12 different orders to do these rotations in, and each triplet of rotated angles is called a set of **Euler angles**. In robotics, the triplet is **yaw**, **pitch** and **roll** (Z-Y-X).

In the NUbots codebase, we denote a rotation matrix from space $b$ to space $a$ as `Rab`.

> There rotation matrices belong to the **special orthogonal group of order 3**. These all have the following properties:

Reviewer: > There rotation matrices

The rotation matrices

$$
\mathbf{R}^{-1} = \mathbf{R}^T
$$

$$
\text{det}(\mathbf{R}) = 1
$$

If we have a vector $\mathbf{r}$ (which goes from point C to point D) in a space $a$ and we want to rotate it from space $a$ to space $b$, we write:

$$

> \mathbf{r}^b = \mathbf{R}^b_a \cdot \mathbf{r}^a

Reviewer: We should add the subscripts to `r` here to tie in the maths notation with the code notation
\mathbf{r}^{b}_{DC} = \mathbf{R}^b_a \cdot \mathbf{r}^{a}_{DC}

$$

This is written in the NUbots codebase as `rCDb = Rba * rCDa`.

> ## Transformations

Reviewer: Homogeneous Transformations

The vector $\mathbf{r}_{AB}^b$ is the vector from B (the origin of space b) to A (the origin of space a) in b space, which would be written in the codebase as `rABb`.

The reason for this is that if we apply the transform $H^b_a$ to some point $\rho^a$, then $\rho^a$ will first be rotated by $R_a^b$, giving $\rho^b$. $\rho^b$ will then be translated by some vector according to the transformation matrix. Since $\rho^b$ is now rotated in the $b$ coordinate system, the translation vector must also be in this $b$ coordinate system. The vector is from B to A because a point (0, 0, 0) in space $a$ should be moved to the origin of $a$ in $b$ space after transformation. Thus, the translation vector is $\mathbf{r}_{AB}^b$.

Transformation matrices left multiply with vectors to get a vector in the new space.

$$
\mathbf{r}^b = \mathbf{H}^b_a \cdot \mathbf{r}^a
$$

This is written in the codebase as `rCDb = Hba * rCDa`, if $\mathbf{r}$ represents a vector from point D to point C.

They also multiply together to get a new transformation matrix.

$$
\mathbf{H}_a^c = \mathbf{H}_b^c \cdot \mathbf{H}^b_a
$$

This is written in the codebase as `Hca = Hcb * Hba`.

> 

Reviewer: We should highlight the difference between points and vectors as they relate to homogeneous transformations.

A point is a vec4 `(x, y, z, 1)`. A point can be rotated and translated.
A vector is a vec4 `(x, y, z, 0)`. A vector can only be rotated.

Author: Do we have any notation for specifying a point? Like how we have rABc for vectors?

Reviewer 2: Our vector notation came from a system where we didn't use homogenous transforms. In that system, there was no need to distinguish as everything was a vector from one point to another. So a vector from the origin of A to some point B is a point because it's referenced to the origin of the space. 99% of the time when we are using our vectors we treat them as points. So those matrices apply rotations and then translations.

Author: That sounds real messy.

If you have a vector `rTWa` and you want the opposite letter order, this is done by negating the vector, `rWTa = -rTWa`. Note that negating does not change the space it is measured in!

If you have two transformation matrices and you would like to multiply them to get a new transformation matrix, the inner two letters should be identical, ie the same space.

```cpp
Eigen::Affine3d Hac = Hab * Hbc
```

These two inner letters are canceled, with the outer two remaining. This also works for rotation matrices.

If you have two vectors in the same space and their inner letters are both the same, then they can be canceled in a similar way to the transformation matrices.

```cpp
Eigen::Vector3d rXZa = rXYa + rYZa;
```

## Summary

**Hab**: Denotes a 3D affine transformation matrix going from space $b$ to space $a$.

**Rab**: Denotes a 3D rotation matrix from space $b$ to space $a$.

**rABb**: Denotes a vector from point $B$ to point $A$ in space $a$.

**vAb**: The velocity of $A$ in space $b$

**rACc = Hcb \* rABb**

**rABd = Rdc \* rABc**

**Hac = Hab \* Hbc**

> **Euler Angles** are of the form Z-Y-X.

Reviewer: Would this look better as a table?

---
section: Projects
chapter: NUbots
title: Mathematics

> description: Detailed explanation of the mathematics used in the NUbots codebase.

Reviewer 3: ```suggestion
description: Detailed explanation of how mathematics is written in the NUbots codebase.
```

slug: /projects/nubots/mathematics
---

> The NUbots codebase uses specific mathematical concepts and notation that is relevant to the field of robotics. This page details these concepts and notations. In the NUbots codebase there are many transforms between spaces. Specific notation is used for these transforms for better readability.

Reviewer 3: ```suggestion
The NUbots codebase uses specific mathematical concepts and notation that are relevant to the field of robotics. In the codebase there are many uses of vectors, matrices, spaces, and transforms between spaces. Specific notation is used for these transforms for better readability. This page details how these concepts and notations are implemented in the codebase.
```

We currently use the [Eigen](https://eigen.tuxfamily.org/) C++ template library for linear algebra in the NUbots codebase.

## Bases

A **basis** is a **maximal** set of **linearly independent** vectors.

> In this context, bases usually span $\mathbb{R}^3$. That is, they cover **three dimensional space**. Given a basis with three vectors $\mathbf{b}_1, \mathbf{b}_2, \mathbf{b}_3$, Any three dimensional vector $\mathbf{u}$ can be written as

Reviewer 3: ```suggestion
In this context, bases usually span $\mathbb{R}^3$. That is, they cover **three dimensional space**. Given a basis with three vectors $\mathbf{b}_1, \mathbf{b}_2, \mathbf{b}_3$, any three dimensional vector $\mathbf{u}$ can be written as
```

$$
\mathbf{u} = a_1 \cdot \mathbf{b}_1 + a_2 \cdot \mathbf{b}_2 + a_3 \cdot \mathbf{b}_3
$$

where $a_1, a_2, a_3 \in \mathbb{R}$. In other words, $a_1, a_2, a_3$ are scalars. These scalars are **unique**.

If the vectors in the basis are mutually perpendicular, then they are **orthogonal**.

An **orthonormal basis (ONB)** is a basis which is orthogonal and each vector has length 1.

An example of an OBN is the **standard basis**

$$
e_1 = (1, 0, 0) \quad
e_2 = (0, 1, 0) \quad
e_3 = (0, 0, 1)
$$

## Vectors

As mentioned in the previous section, any vector $\mathbf{r}$ can be expressed as a linear combination of the elements of a basis

$$
\mathbf{r}^a = r_1^a \mathbf{a}_1 + r_2^a \mathbf{a}_2 + r_3^a \mathbf{a}_3
$$

The scalars $r_1^a, r_2^a, r_3^a$ are the **coordinates** for the vector $\mathbf{r}$ in the basis $a$.

In the NUbots codebase, vectors are written as `rABb` which denotes a vector from point $B$ to point $A$ in the basis (or **space**) $b$. These are column matrices of the coordinates of the vectors

$$
\mathbf{r}^a = \begin{bmatrix}
r_1^a \\
r_2^a \\
r_3^a \end{bmatrix}
$$

> To make computations such as **inner product** or **cross product** with coordinates, the vectors must be relative to the **same basis**.

Author: technically not the same thing but in this case I guess they are :P

When drawing a vector $\mathbf{r}_{AB}^b$, that is a vector from $B$ to $A$ in the space $b$, the **tail** is at $B$ and the **head** is at $A$. **Vector addition** requires that the head of one vector is the same as the tail of the other. This is demonstrated [here](http://mathworld.wolfram.com/VectorAddition.html). We have that

$$
\mathbf{r}^b_{AC} = \mathbf{r}^b_{AB} + \mathbf{r}^b_{BC} = \mathbf{r}^b_{BC} + \mathbf{r}^b_{AB}
$$

since vector addition is **commutative**. This is written in code as

```cpp
Eigen::Vector3d rACb = rABb + rBCb;
```

## Rotations

To describe the motion of the robot, we need a **reference frame** and a **coordinate system**.

The reference frame is the **physical rigid body** and the coordinate system is a **mathematical concept**, which is represented as a **basis**.

> NUbots use a **right-handed coordinate system** where the x-axis is **forward**, the y-axis is to the **left** and the z-axis is **up**.

Reviewer 3: Probably too much to ask but a diagram/illustration with hand would be great here.

Reviewer 3: Something like the one on Wikipedia: https://en.wikipedia.org/wiki/Right-hand_rule

$$
\mathbf{H}_a^c = \mathbf{H}_b^c \cdot \mathbf{H}^b_a
$$

This is written in the codebase as

```cpp
Eigen::Affine3d Hca = Hcb * Hba;
```

## Velocity

Velocity of some object is represented relative to some space. If we have the **velocity** of object A moving **relative** to space b, then we write this in the codebase as `vAb`.

Velocities can only be **rotated**, **not translated**. They should **not** be multiplied by a homogeneous transformation matrix, but they can be multiplied by a **rotation matrix**.

If a velocity is multiplied with **time**, it becomes a **vector**.

```cpp
Eigen::Vector3d rA_tAb = vAb * t;
```

## Examples

If you have a transformation matrix `Htw` (**world** to **torso** transform) and you want the matrix `Hwt` (**torso** to **world** transform), this is can be achieved by calling `Htw.inverse()`. The inverse function **swaps** the letter order. The returned inverse matrix is not **mutable**.

If you have a vector `rTWf` (**world** to **torso** in **foot** space) and you want the opposite letter order (**torso** to **world** in **foot** space), this is done by negating the vector, `rWTf = -rTWf`.

> <Alert type='warning'>

Reviewer 3: Can't decide whether this is more of an informational alert than a warning.

$$
\mathbf{r}^a = \begin{bmatrix}
r_1^a \\
r_2^a \\
r_3^a \end{bmatrix}
$$

To make computations such as **inner product** (also called **dot product**) or **cross product** with coordinates, the vectors must be relative to the **same basis**.

When drawing a vector $\mathbf{r}_{AB}^b$, that is a vector from $B$ to $A$ in the space $b$, the **tail** is at $B$ and the **head** is at $A$. **Vector addition** requires that the head of one vector is the same as the tail of the other. This is demonstrated [here](http://mathworld.wolfram.com/VectorAddition.html). We have that

$$
\mathbf{r}^b_{AC} = \mathbf{r}^b_{AB} + \mathbf{r}^b_{BC} = \mathbf{r}^b_{BC} + \mathbf{r}^b_{AB}
$$

since vector addition is **commutative**. This is written in code as

```cpp
Eigen::Vector3d rACb = rABb + rBCb;
```

## Rotations

To describe the motion of the robot, we need a **reference frame** and a **coordinate system**.

The reference frame is the **physical rigid body** and the coordinate system is a **mathematical concept**, which is represented as a **basis**.

NUbots use a **right-handed coordinate system** where the x-axis is **forward**, the y-axis is to the **left** and the z-axis is **up**.

> ![A right hand with the index finger pointed forwards and labelled 'a', the middle finger pointing to the left and labelled 'b', and the thumb pointing upwards and labelled 'a x b'](./images/right-hand-rule.svg 'Right-Hand Rule. Image from Wikimedia Commons https://en.wikipedia.org/wiki/File:Right_hand_rule_cross_product.svg')

Reviewer 3: Would it make sense to change the image to have `x`, `y` and `z` like in the text or change the text to match the image?

Author: Changed image have x, y, z.
